src/mds/CDir.cc

```diff
@@ -1815,8 +1810,18 @@ void CDir::_omap_fetched(bufferlist& hdrbl, map<string, bufferlist>& omap,
       cache->mds->clog->warn() << "Corrupt dentry '" << dname << "' in "
                                   "dir frag " << dirfrag() << ": "
                                << err;
-      go_bad();
-      return;
+
+      // Remember that this dentry is damaged.  Subsequent operations
+      // that try to act directly on it will get their EIOs, but this
+      // dirfrag as a whole will continue to look okay (minus the
+      // mysteriously-missing dentry)
+      go_bad_dentry(last, dname);
+
+      // Anyone who was WAIT_DENTRY for this guy will get kicked
+      // to RetryRequest, and hit the DamageTable-interrogating path.
+      // Stats will now be bogus because we will think we're complete,
+      // but have 1 or more missing dentries.
+      continue;
```

Comment:
I'm not sure this is quite sufficient. It's good to keep reading stuff, but don't we also need to avoid marking the directory complete? That's used as a marker for more than just lookup — scrubbing for instance (I'm not sure what the behavior should be there when we know part of a dirfrag is damaged...), split/merge decisions, which we probably don't want to enter when dentries are damaged, etc.

Of course, there's also other stuff we *do* want to do even if damaged, so I'm not sure what the answer is here. We may need to add another dimension to all those decisions points; please comment.